Fix sly change directory

Ambrevar opened the pull request:
A few "cd"-related issues here.

Changing directory should happen both in the Lisp image and in the various mREPLs connected to that image.

I don't think it's OK to call mrepl code in sly.el, so `sly-change-directory` is probably not right as I've patched it.

Maybe move sly-change-directory to contrib/sly-mrepl?

Ambrevar commented:
Maybe better: move `sly-cd` to contrib/sly-mrepl, have sly-cd call sly-mrepl-set-directory and loop over all REPLs in this last function instead of sly-set-directory.

joaotavora commented on sly.el:
```diff
@@ -4410,10 +4410,27 @@ The functions are called with the new (absolute) directory.")
+                 (buffer-list))))
     (prog1 (sly-eval `(slynk:set-default-directory
-                       ,(sly-to-lisp-filename dir)))
+                       (uiop:parse-native-namestring
```
Can't rely on UIOP.  Slynk should run on older Lisps that don't have this.

joaotavora reviewed (commented):
I like the possibility and ease of having each REPL have a different current directory.  As to the special characters thing, you have to find something that doesn't need UIOP.

Ambrevar commented:
> I like the possibility and ease of having each REPL have a different current directory.  
Note that in this case, (uiop:getcwd) will be different from Emacs'
`default-directory`.
Do we really want to confuse our users?

I use may REPLs connected to the same inferior and I find it confusing
when I change directory in one, but the change is not forwarded to the
other REPL.

What about making this an option then?  I think it would be useful and
it does cost much :)

Ambrevar commented:
João Távora <notifications@github.com> writes:

> Can't rely on UIOP.  Slynk should run on older Lisps that don't have this.

OK, but I don't think there is a way to do this portable.  Here is the
UIOP code:

```
(defun parse-native-namestring (string &rest constraints &key ensure-directory &allow-other-keys)
    "From a native namestring suitable for use by the operating system, return
a CL pathname satisfying all the specified constraints as per ENSURE-PATHNAME"
    (check-type string (or string null))
    (let* ((pathname
             (when string
               (with-pathname-defaults ()
                 #+clozure (ccl:native-to-pathname string)
                 #+cmucl (uiop/os::parse-unix-namestring* string)
                 #+sbcl (sb-ext:parse-native-namestring string)
                 #+scl (lisp::parse-unix-namestring string)
                 #-(or clozure cmucl sbcl scl)
                 (os-cond
                  ((os-unix-p) (parse-unix-namestring string :ensure-directory ensure-directory))
                  (t (parse-namestring string))))))
           (pathname
             (if ensure-directory
                 (and pathname (ensure-directory-pathname pathname))
                 pathname)))
      (apply 'ensure-pathname pathname constraints)))
```

joaotavora commented:
> OK, but I don't think there is a way to do this portable. 

This would have to go into a `definterface` and `defimplementation` pair like everyone else.  See `slynk-backend.lisp`

> What about making this an option then?  I think it would be useful and it does cost much :)

You can make this a contrib.

> Note that in this case, (uiop:getcwd) will be different from Emacs' `default-directory`.
> Do we really want to confuse our users?

I've never heard of this report.  I use `C-c ~` to synch package and directory, and it works fine.  
Also, there's no one single Emacs's `default-directory`.  It's a per-buffer variable.  What buffer 
are you referring to?

I.e. I need a recipe from scratch so I can understand the potential problem before diving
into solutions.

Ambrevar commented:
João Távora <notifications@github.com> writes:

> This would have to go into a `definterface` and `defimplementation` pair like everyone else.  See `slynk-backend.lisp`

Will do.

>> What about making this an option then?  I think it would be useful and it does cost much :)
>
> You can make this a contrib.

To be clear, you mean in a separate repository, not in the `contrib/`
folder, right?

> I've never heard of this report.  I use `C-c ~` to synch package and directory, and it works fine.  
> Also, there's no one single Emacs's `default-directory`.  It's a per-buffer variable.  What buffer 
> are you referring to?
>
> I.e. I need a recipe from scratch so I can understand the potential problem before diving
> into solutions.

- M-x sly
  ; (uiop:getcwd) => /path/bar
  ; default-directory => /path/bar (let's assume it's the same as (uiop:getcwd))

- M-x sly-mrepl-new RET foo RET
  ; (uiop:getcwd) => /path/bar
  ; default-directory => /path/bar 

- M-x sly-cd RET /path/qux RET
  ; (uiop:getcwd) => /path/qux
  ; default-directory => /path/bar 

- M-x switch-to-buffer RET *sly-mrepl* RET ; the first REPL
  ; (uiop:getcwd) => /path/qux 
  ; default-directory => /path/bar 

- ,cd RET /path/frobnicator RET
  ; (uiop:getcwd) => /path/frobnicator
  ; default-directory => /path/frobnicator

- M-x switch-to-buffer RET *sly-mrepl (foo)* RET ; the second REPL
  ; (uiop:getcwd) => /path/frobnicator
  ; default-directory => /path/bar

,cd and sly-cd are not consistent.

- It's confusing to sly-cd only to realize that the Emacs'
  default-directory of the current buffer is still the old path.

- It's confusing to ,cd which changes the inferior uiop:getcwd _and_ the
  Emacs default-directory, but not that of other mrepls.

I hope this helps understand the problem :)

joaotavora commented:
> I hope this helps understand the problem :)

It's better but it's still confusing.  I need to know which buffer you are at all times.  So in the third step are you in the second REPL buffer?  Better post the minimal sequence of commands that leads the inconsitency.  I can follow them like if I was a robot.

>  ; default-directory => /path/bar

Is this `C-h v d e f a u l t - d i r e c t o r y RET` or equivalent, like `C-:` with. 

But in appears that making `sly-change-directory` affect `default-directory` and that would fix the discrepancy.  At least in the REPL.  `sly-change-directory-hooks` seems like a decent place to do that. Anyway please try again with an idiot-proof recipe. (me being the idiot here, of course ;-) )

Ambrevar commented:
About the special char escaping issue: turns out Slynk already has slynk-backend:filename-to-pathname that does the job!
I've sent a final fix for this issue in #381.

Ambrevar commented:
> So in the third step are you in the second REPL buffer? 

Yes.

I'll make a script if that can help.

joaotavora commented:
> ,cd and sly-cd are not consistent.

This seems to be the problem. I'll have a look.

As to the recipe.  No script, just an unambiguous recipe would help.  But I think I see the problem.

Ambrevar commented:
Too late ;)

```elisp
(cl-flet ((report
           ()
           (message "==> SLYDEBUG: %s\nLisp: %s\nEmacs: %s"
                    (current-buffer)
                    (sly-eval '(cl:namestring (uiop:getcwd)))
                    default-directory))
          (sly-mrepl-set-directory/unattended
           (directory)
           "Like `sly-mrepl-set-directory' but takes an argument instead of prompting."
           (sly-mrepl--save-and-copy-for-repl
            `(slynk:set-default-directory
              (slynk-backend:filename-to-pathname
               ,directory))
            :before (format "Setting directory to %s" directory))))
  (cd "~")
  (let ((repl1 (sly)))
    (sit-for 2)
    (report)
    (let ((repl2 (sly-mrepl-new (sly-connection) "NEW")))
      (report)
      (sly-cd "/tmp")
      (report)
      (switch-to-buffer repl1)
      (report)
      (message "[sly] Changing Lisp getcwd")
      (sly-mrepl-set-directory/unattended "/")
      (report)
      (switch-to-buffer repl2)
      (report)
      nil)))
```

Ambrevar commented:
I've addressed all the issues.
I'm no longer switching directories for all REPL: indeed, each REPL can have its own default directory.
This change introduced bug #382.

I believe we are all set, this should be good to merge.

joaotavora commented:
I'm sorry, but this is not good to merge: you can't call `sly-mrepl` code in `sly.el`.

Ambrevar commented:
Oh, missed that.

My intention was to notify the user about the directory change on
`sly-cd'.

Any suggestion on how to do this?
What about moving sly-cd to sly-mrepl?
Alternatively, we could define sly-mrepl-cd?

joaotavora commented:
Maybe some of those work.  Try to find the one that changes less code, respects this separation, and doesn't break any existing use cases. I'm sorry, I don't have time to dive into this (or any of the other issues) right now.

Ambrevar commented:
Done.  The main changes:

- `sly-cd` is now moved to `sly-mrepl.el`: this makes sense, since it changes the directory of the REPL.
  `sly-cd` wasn't used anywhere.
- `sly-mrepl-set-directory` leverages `sly-change-directory` thanks to a new optional argument to the latter.

This should be completely backward compatible (unless sly-mrepl.el is not loaded and the user tries to use sly-cd, which should never happen...?).

joaotavora commented:
I'm closing this in favour or the fix I pushed for #382.  

But if you think there are some outstanding things I didn't address, please do comment and we may reopen.

Ambrevar commented:
Looks like this doesn't address the initial issue that it fails on wild
cards (e.g. `*`, `[`).

Either the inputted dir must be wrapped in
`slynk-backend:filename-to-pathname`, or `slynk:set-default-directory'
must turn the pathname to a parse native pathname.

joaotavora commented:
Can you explain the "wild card" problem again, please?

Ambrevar commented:
Try running any of the dir switching command on a directory that
contains a "[" character, it will fail.

Ambrevar commented:
(This is because in Common Lisp, pathnames interpret wildcards by
default.)

joaotavora commented:
Then I guess we need some kind of escaping.  Can you point to the code that does that?  This PR has a lot of mixed things and I don't have time do search.

joaotavora commented:
OK.  Ive seen the UIOP code that parses "native namestrings".  To do such a thing in SLY, it needs a new interface in `slynk-backend.lisp`, and implementations for all Lisps, which might be too much here. I'd rather do the conversion on the Emacs-lisp side.

Ambrevar commented on sly.el:
```diff
@@ -4407,21 +4407,17 @@ in Lisp when committed with \\[sly-edit-value-commit]."
-                       ,(sly-to-lisp-filename dir)))
+    (prog1 (funcall (or evaluator 'sly-eval)
+                    `(slynk:set-default-directory
+                      (slynk-backend:filename-to-pathname
```
It's just this one line that fixes it: `slynk-backend:filename-to-pathname`.
But this only works if the other cd functions call `sly-change-directory`.

Ambrevar commented:
> OK. Ive seen the UIOP code that parses "native namestrings". To do such a thing in SLY, it needs a new interface in slynk-backend.lisp

It's already implemented as `slynk-backend:filename-to-pathname`.
(In fact, we already had this very same discussion, haha:p https://github.com/joaotavora/sly/pull/379#issuecomment-730452644)

joaotavora commented:
> In fact, we already had this very same discussion, haha:p

Yeah, you open so many issues that I get lost sorry.  Maybe we should enable GitHub "conversations", and then you can make threads at will or raise random issues in one of those threads.  We'll make issues if they are deemed worth it.

I'll see how I can leverage `slynk-backend:filename-to-pathname`

joaotavora commented:
There are already discussions here: https://github.com/joaotavora/sly/discussions

I propose we start conversing there.  Then we agree whether to make issues or not.

Ambrevar commented:
Oh, just saw the new discussions tab!
Nice!

Sorry about the many issues, I've been very busy hacking SLY these weeks
for my FOSDEM presentation.

Things should calm down after that! :)

joaotavora commented:
By the way, are you sure this doesn't fix it?
```diff
diff --git a/contrib/sly-mrepl.el b/contrib/sly-mrepl.el
index 640d9e48..be22890f 100644
--- a/contrib/sly-mrepl.el
+++ b/contrib/sly-mrepl.el
@@ -1314,7 +1314,7 @@ When setting this variable outside of the Customize interface,
   (let ((directory (read-directory-name "New directory: "
                                         default-directory nil t)))
     (sly-mrepl--eval-for-repl
-     `(slynk:set-default-directory ,directory))
+     `(slynk:set-default-directory ,(sly-to-lisp-filename directory)))
     (sly-mrepl--insert-note (format "Setting directory to %s" directory))
     (cd directory)))
 
```

joaotavora commented:
Also, thanks for promoting SLY in your talks.  I appreciate it.

Ambrevar commented:
I confirm that it does not work, while 

> ```diff
> diff --git a/contrib/sly-mrepl.el b/contrib/sly-mrepl.el
> index 640d9e48..be22890f 100644
> --- a/contrib/sly-mrepl.el
> +++ b/contrib/sly-mrepl.el
> @@ -1314,7 +1314,7 @@ When setting this variable outside of the Customize interface,
>    (let ((directory (read-directory-name "New directory: "
>                                          default-directory nil t)))
>      (sly-mrepl--eval-for-repl
> -     `(slynk:set-default-directory ,directory))
> +     `(slynk:set-default-directory (slynk-backend:filename-to-pathname ,directory)))
>      (sly-mrepl--insert-note (format "Setting directory to %s" directory))
>      (cd directory)))
>  
> ```

works.

But this patch alone would not fix sly-cd.

Ambrevar commented:
Thanks for making SLY! :)

Here is the event and the future video: https://fosdem.org/2021/schedule/event/lisprepl/

joaotavora commented:
> But this patch alone would not fix sly-cd.

Then make a patch that also uses this function in `sly-cd`. I think both functions should work similarly, probably leveragin both `sly-to-lisp-filename` and `slynk-backend:filename-to-pathname`.  Maybe `slynk:set-default-directory` should use use `slynk-backend:filename-to-pathname`, then both users of it would be served.

joaotavora commented:
Now both directory switchers use `slynk-backend:filename-to-pathname`

Ambrevar commented:
This is perfect, thank you very much!
